### PR TITLE
S22 North Pole: Handle missing cases in MP/defense update

### PR DESCRIPTION
**Pull request by Slayer95**

### CODE REVIEW

**Comment by Slayer95** [comment]
Funny note: After attempting to "pass-by fix" undoing recalls of dimensional gates for many hours, I have become enlightened on Edison's "a thousand ways not to make a light bulb".

**Comment by ProditorMagnus** [comment]
Did you consider unit placed?

**Comment by Slayer95** [comment]
> Did you consider unit placed?

Not really. But according to the wiki, it should be listened to only sparingly, right?

**Comment by ProditorMagnus** [comment]
It is for cases where you want to be sure something is done for unit no matter how it appears.

**Comment by Slayer95** [comment]
Does it include advancements? Such units were already in place, kinda. I am not familiar with the extent of the event.

**Comment by Slayer95** [comment]
Alright. Switching to ``unit placed`` uncovered a few bugs. Among other things, it will now correctly update the speed of Dharma'Rashti if she spawns.

**Comment by Slayer95** [comment]
Could you review again, please? @ProditorMagnus

File: scenarios/22_North_Pole.cfg [hunk 1]
```diff
@@ -23,6 +23,85 @@
         [/lua]
     [/event]
 
+    #define IMPROVE_STATS_ON_FROZEN UNIT_FILTER
```
Comment (ProditorMagnus) [inline scenarios/22_North_Pole.cfg]: #define in scenario should be #undef in same file

Comment (ProditorMagnus) [reply scenarios/22_North_Pole.cfg]: preprocessor directives should not have whitespace before them

File: scenarios/22_North_Pole.cfg [hunk 2]
```diff
@@ -23,6 +23,85 @@
...
+                    {VARIABLE def_bonus 60}
+                    {IF_VAR unit.type equals EoMa_Um (
+                        [then]
+                            {VARIABLE def_bonus 70}
+                        [/then]
+                    )}
+                    {IMPROVE_STATS_ON_FROZEN (id=$unit.id) DEF=$def_bonus}
+                    {CLEAR_VARIABLE def_bonus}
+                )}
+                # Cache unit type ("unit placed" is spammed on [harm_unit], and who knows elsewhen.)
```
Comment (ProditorMagnus) [inline scenarios/22_North_Pole.cfg]: has_frozen_def_bonus would be more informative than unit_type_when_bonus_set. Unit placed is indeed supposed to happen multiple times for single unit.

Comment (ProditorMagnus) [reply scenarios/22_North_Pole.cfg]: ideally you have it as part of object filter https://github.com/ProditorMagnus/OrociaRandomMod/blob/9ae304204317ff1299609da22443c695690ad298/lua/helper_functions.lua#L84-L94

File: scenarios/22_North_Pole.cfg [hunk 3]
```diff
@@ -23,6 +23,85 @@
...
+                {MODIFY_UNIT id=$unit.id variables.unit_type_when_bonus_set $unit.type}
+            [/then]
+        )}
+        {CLEAR_VARIABLE unit_type_when_bonus_set}
+
+        [allow_undo]
+        [/allow_undo]
+        [on_undo]
+            # invalidate cache
+            {MODIFY_UNIT id=$unit.id unit.variables.unit_type_when_bonus_set $empty}
```
Comment (ProditorMagnus) [inline scenarios/22_North_Pole.cfg]: why this is needed?

File: scenarios/22_North_Pole.cfg [hunk 4]
```diff
@@ -23,6 +23,85 @@
...
+                    {IMPROVE_STATS_ON_FROZEN (id=$unit.id) DEF=$def_bonus}
+                    {CLEAR_VARIABLE def_bonus}
+                )}
+                # Cache unit type ("unit placed" is spammed on [harm_unit], and who knows elsewhen.)
+                {MODIFY_UNIT id=$unit.id variables.unit_type_when_bonus_set $unit.type}
+            [/then]
+        )}
+        {CLEAR_VARIABLE unit_type_when_bonus_set}
+
+        [allow_undo]
```
Comment (ProditorMagnus) [inline scenarios/22_North_Pole.cfg:97]: changing unit variables is less dangerous oos change than changing unit defense. If undo is legal and does not cause oos for wrong def, there is no benefit to remove the flag of unit having object

File: scenarios/22_North_Pole.cfg (hunk 2 above)
Comment (Slayer95) [reply scenarios/22_North_Pole.cfg]: > has_frozen_def_bonus would be more informative than unit_type_when_bonus_set.

The problem is that some units (e.g. Fire Elemental) must lose its "bonus" (in fact, a replace) when leveling up, so we gotta keep track of the unit type in order to recheck it and reapply the bonus if pertinent.

File: scenarios/22_North_Pole.cfg (hunk 3 above)
Comment (Slayer95) [reply scenarios/22_North_Pole.cfg]: Just for completeness. Unless I am missing something, it shouldn't be good to change the unit state on an action, but keep the changes when undoing it.

File: scenarios/22_North_Pole.cfg (hunk 1 above)
Comment (Slayer95) [reply scenarios/22_North_Pole.cfg]: Noted. Will fix.

File: scenarios/22_North_Pole.cfg (hunk 4 above)
Comment (Slayer95) [reply scenarios/22_North_Pole.cfg:97]: I see your point. Will test without ``[on_undo]``

File: scenarios/22_North_Pole.cfg (hunk 2 above)
Comment (Slayer95) [reply scenarios/22_North_Pole.cfg]: Note that ``has_frozen_def_bonus ↔ unit_type_when_bonus_set ∈ flying_unit_types``. Therefore, the information provided by ``has_frozen_def_bonus`` is also readily available from ``unit_type_when_bonus_set``.

File: scenarios/22_North_Pole.cfg (hunk 4 above)
Comment (ProditorMagnus) [reply scenarios/22_North_Pole.cfg:97]: If you have unit placed with [filter] to only act when unit does not already have object, I would expect it work good enough even without allowing undo.

File: scenarios/22_North_Pole.cfg (hunk 2 above)
Comment (ProditorMagnus) [reply scenarios/22_North_Pole.cfg]: You cant add filter inside [effect] for this use case?

Comment (Slayer95) [reply scenarios/22_North_Pole.cfg]: > You cant add filter inside [effect] for this use case?

I thought ``[filter]`` was a tag for ``[object]`` only. It's great that it works with ``[effect]``. I'd better do that!

**Comment by Slayer95** [comment]
Ok. I found ``[effect][filter]`` has more nuances than just a Fire Elemental leveling up, and that ``[filter_wml]`` actually has a fast path for ``[variables]``, so I went with ``[event][filter]``.

File: scenarios/22_North_Pole.cfg [hunk 5]
```diff
@@ -327,46 +370,8 @@
...
-        name=recruit,recall,post advance
-        first_time_only=no
-
-        [filter]
-            side=1
-            type=EoMa_Dimensional_Gate,EoMa_Dimensional_Gate_II,EoMa_Jinn,EoMa_Carpet_Rider,EoMa_Great_Jinn,EoMa_Wonderful_Jinn,EoMa_Mystical_Jinn,EoMa_Efreet,EoMa_Great_Efreeti,EoMa_Carpet_Master,EoMa_Air_Elemental,EoMa_Air_Avatar,EoMa_Air_God,EoMa_Fire_Avatar,EoMa_Fire_God
-        [/filter]
-
-        {MODIFY_UNIT (x,y=$x1,$y1) movement_costs.frozen 1}
+        [allow_undo]
```
Comment (ProditorMagnus) [inline scenarios/22_North_Pole.cfg:373]: I am not convinced all the events where you allow undo are actually safe to undo

Comment (Slayer95) [reply scenarios/22_North_Pole.cfg:373]: Well, ``[allow_undo]`` has no effect for events for which it's known that cannot be undone (e.g. ``post advance``). I've played around with undoing actions in this scenario plenty, and I have only managed to trigger OOS when attempting to undo a recall event that called ``{ADVANCE_UNIT}``, so these should be safe.

**Pull request merged**